pkg/scheduler/algorithm/priorities/requested_to_capacity_ratio.go

```diff
@@ -0,0 +1,145 @@
+			return FunctionShape{}, fmt.Errorf("values in x must not be less than %d. x[%d]==%d", minX, i, x[i])
+		}
+		if x[i] > maxX {
+			return FunctionShape{}, fmt.Errorf("values in x must not be greater than %d. x[%d]==%d", maxX, i, x[i])
+		}
+		if y[i] < minY {
+			return FunctionShape{}, fmt.Errorf("values in y must not be less than %d. y[%d]==%d", minY, i, y[i])
+		}
+		if y[i] > maxY {
+			return FunctionShape{}, fmt.Errorf("values in y must not be greater than %d. y[%d]==%d", maxY, i, y[i])
+		}
+	}
+
+	return FunctionShape{
+		x: x,
+		y: y,
+	}, nil
+}
+
+// RequestedToCapacityRatioResourceAllocationPriorityDefault creates a requestedToCapacity based
+// ResourceAllocationPriority using default resource scoring function shape.
+// The default function assigns 1.0 to resource when all capacity is available
+// and 0.0 when requested amount is equal to capacity.
+func RequestedToCapacityRatioResourceAllocationPriorityDefault() *ResourceAllocationPriority {
+	return RequestedToCapacityRatioResourceAllocationPriority(defaultFunctionShape)
+}
+
+// RequestedToCapacityRatioResourceAllocationPriority creates a requestedToCapacity based
+// ResourceAllocationPriority using provided resource scoring function shape.
+func RequestedToCapacityRatioResourceAllocationPriority(scoringFunctionShape FunctionShape) *ResourceAllocationPriority {
+	return &ResourceAllocationPriority{"RequestedToCapacityRatioResourceAllocationPriority", buildRequestedToCapacityRatioScorerFunction(scoringFunctionShape)}
+}
+
+func buildRequestedToCapacityRatioScorerFunction(scoringFunctionShape FunctionShape) func(*schedulercache.Resource, *schedulercache.Resource, bool, int, int) int64 {
+	rawScoringFunction := buildBrokenLinearFunction(scoringFunctionShape)
+
+	resourceScoringFunction := func(requested, capacity int64) int64 {
+		if capacity == 0 || requested > capacity {
+			return rawScoringFunction(maxX)
+		}
+
+		return rawScoringFunction(maxX - (capacity-requested)*maxX/capacity)
+	}
+
+	return func(requested, allocable *schedulercache.Resource, includeVolumes bool, requestedVolumes int, allocatableVolumes int) int64 {
+		cpuScore := resourceScoringFunction(requested.MilliCPU, allocable.MilliCPU)
+		memoryScore := resourceScoringFunction(requested.Memory, allocable.Memory)
+		return (cpuScore + memoryScore) / 2
+	}
+}
+
+// Creates a function which is built using linear segments
+// shape.x slice represents points on x axis where different segments meet
+// shape.y represents function values at meeting points
+// both shape.x and shape.y have same length (n)
+//
+// function f(p) is defined as:
+//   y[0] for p < x[0]
+//   y[i] for p == x[i]
+//   y[n-1] for p > x[n-1]
+// and linear between points (p < x[i])
+func buildBrokenLinearFunction(shape FunctionShape) func(int64) int64 {
+	if len(shape.x) != len(shape.y) {
+		glog.Fatalf("invalid argument; len(shape.x)==%d, len(shape.y)==%d", shape.x, shape.y)
+	}
+	n := len(shape.x)
+	x := make([]int64, n)
+	copy(x, shape.x)
```

Comment:
The returned function is not changing the passed values. Also the shape is not expected to change at run-time. So, I think copying is not needed here.

Reply:
This is strictly defensive. By having copy here we are not making the assumption what caller of buildBrokenLinearFunction does with parameter afterwards. IMO it is sane approach. As it is not performance critical I would leave copy here.

Reply:
That's fine, but please add a comment and explain what you said in a comment so that future developers don't get confused.

Reply:
I moved copying to `NewFunctionShape`, and dropped it from here. It feels not ideal. Aat least to me, as I am used to having immutable data structures where possible, and Java conveniences of guava like `ImmutableList.copyOf` which may not do the copy if parametr is already an `ImmutableList`. Yet maybe I have to get used to it as raw arrays are commonly passed around in golang.